In [30]:
import pandas as pd

Loaded the Dataset

In [31]:
df=pd.read_csv(r"D:\Enterprise_ISTM_AI\data\raw\incident_event_log.csv")

Inspecting the values

In [32]:
df.head(5)

,number,incident_state,active,reassignment_count,reopen_count,sys_mod_count,made_sla,caller_id,opened_by,opened_at,...,u_priority_confirmation,notify,problem_id,rfc,vendor,caused_by,closed_code,resolved_by,resolved_at,closed_at
0,INC0000045,New,True,0,0,0,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
1,INC0000045,Resolved,True,0,0,2,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
2,INC0000045,Resolved,True,0,0,3,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
3,INC0000045,Closed,False,0,0,4,True,Caller 2403,Opened by 8,29/2/2016 01:16,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 149,29/2/2016 11:29,5/3/2016 12:00
4,INC0000047,New,True,0,0,0,True,Caller 2403,Opened by 397,29/2/2016 04:40,...,False,Do Not Notify,?,?,?,?,code 5,Resolved by 81,1/3/2016 09:52,6/3/2016 10:00


Size of the dataset

In [33]:
df.shape

(141712, 36)

Column names

In [34]:
df.columns

Index(['number', 'incident_state', 'active', 'reassignment_count',
       'reopen_count', 'sys_mod_count', 'made_sla', 'caller_id', 'opened_by',
       'opened_at', 'sys_created_by', 'sys_created_at', 'sys_updated_by',
       'sys_updated_at', 'contact_type', 'location', 'category', 'subcategory',
       'u_symptom', 'cmdb_ci', 'impact', 'urgency', 'priority',
       'assignment_group', 'assigned_to', 'knowledge',
       'u_priority_confirmation', 'notify', 'problem_id', 'rfc', 'vendor',
       'caused_by', 'closed_code', 'resolved_by', 'resolved_at', 'closed_at'],
      dtype='object')

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 141712 entries, 0 to 141711
Data columns (total 36 columns):
 #   Column                   Non-Null Count   Dtype 
---  ------                   --------------   ----- 
 0   number                   141712 non-null  object
 1   incident_state           141712 non-null  object
 2   active                   141712 non-null  bool  
 3   reassignment_count       141712 non-null  int64 
 4   reopen_count             141712 non-null  int64 
 5   sys_mod_count            141712 non-null  int64 
 6   made_sla                 141712 non-null  bool  
 7   caller_id                141712 non-null  object
 8   opened_by                141712 non-null  object
 9   opened_at                141712 non-null  object
 10  sys_created_by           141712 non-null  object
 11  sys_created_at           141712 non-null  object
 12  sys_updated_by           141712 non-null  object
 13  sys_updated_at           141712 non-null  object
 14  contact_type        

Checking the data types

In [36]:
df.dtypes

number                     object
incident_state             object
active                       bool
reassignment_count          int64
reopen_count                int64
sys_mod_count               int64
made_sla                     bool
caller_id                  object
opened_by                  object
opened_at                  object
sys_created_by             object
sys_created_at             object
sys_updated_by             object
sys_updated_at             object
contact_type               object
location                   object
category                   object
subcategory                object
u_symptom                  object
cmdb_ci                    object
impact                     object
urgency                    object
priority                   object
assignment_group           object
assigned_to                object
knowledge                    bool
u_priority_confirmation      bool
notify                     object
problem_id                 object
rfc           

In [37]:
df.describe()

,reassignment_count,reopen_count,sys_mod_count
count,141712.000000,141712.000000,141712.000000
mean,1.104197,0.021918,5.080946
std,1.734673,0.207302,7.680652
min,0.000000,0.000000,0.000000
25%,0.000000,0.000000,1.000000
50%,1.000000,0.000000,3.000000
75%,1.000000,0.000000,6.000000
max,27.000000,8.000000,129.000000


Checking column Variables which are not numerical

In [38]:
cols=df.select_dtypes(include='object').columns
print(cols)

Index(['number', 'incident_state', 'caller_id', 'opened_by', 'opened_at',
       'sys_created_by', 'sys_created_at', 'sys_updated_by', 'sys_updated_at',
       'contact_type', 'location', 'category', 'subcategory', 'u_symptom',
       'cmdb_ci', 'impact', 'urgency', 'priority', 'assignment_group',
       'assigned_to', 'notify', 'problem_id', 'rfc', 'vendor', 'caused_by',
       'closed_code', 'resolved_by', 'resolved_at', 'closed_at'],
      dtype='object')


Converting ? to NAN VALUES

In [39]:
df.replace('?',pd.NA,inplace=True)

In [40]:
df.isnull().sum()

number                          0
incident_state                  0
active                          0
reassignment_count              0
reopen_count                    0
sys_mod_count                   0
made_sla                        0
caller_id                      29
opened_by                    4835
opened_at                       0
sys_created_by              53076
sys_created_at              53076
sys_updated_by                  0
sys_updated_at                  0
contact_type                    0
location                       76
category                       78
subcategory                   111
u_symptom                   32964
cmdb_ci                    141267
impact                          0
urgency                         0
priority                        0
assignment_group            14213
assigned_to                 27496
knowledge                       0
u_priority_confirmation         0
notify                          0
problem_id                 139417
rfc           

Changing the date time values format to datetime

In [41]:
date_columns=['opened_at','sys_created_at','sys_updated_at','resolved_at']
for col in date_columns:
    df[col]=pd.to_datetime(df[col],errors='coerce')

C:\Users\bharg\AppData\Local\Temp\ipykernel_7344\2884995440.py:3: UserWarning: Parsing dates in %d/%m/%Y %H:%M format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df[col]=pd.to_datetime(df[col],errors='coerce')
C:\Users\bharg\AppData\Local\Temp\ipykernel_7344\2884995440.py:3: UserWarning: Parsing dates in %d/%m/%Y %H:%M format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df[col]=pd.to_datetime(df[col],errors='coerce')
C:\Users\bharg\AppData\Local\Temp\ipykernel_7344\2884995440.py:3: UserWarning: Parsing dates in %d/%m/%Y %H:%M format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df[col]=pd.to_datetime(df[col],errors='coerce')
C:\Users\bharg\AppData\Local\Temp\ipykernel_7344\2884995440.py:3: UserWarning: Parsing dates in %d/%m/%Y %H:%M format when dayfirst=False (the default)

In [42]:
df[date_columns].dtypes

opened_at         datetime64[ns]
sys_created_at    datetime64[ns]
sys_updated_at    datetime64[ns]
resolved_at       datetime64[ns]
dtype: object

In [43]:
print(df.duplicated().sum())

0


In [44]:
df.drop(columns=['cmdb_ci'],axis=1,inplace=True,errors='ignore')

In [45]:
#deleting closed_at column
df.drop(columns=['closed_at'],axis=1,inplace=True,errors='ignore')

In [46]:
#creating a temporary missing dataframe
missing=pd.DataFrame({'missing':df.isnull().sum(),
                      'missing_percentage':df.isnull().mean()*100})
missing=missing.sort_values(by='missing_percentage',ascending=False)
print(missing)

                         missing  missing_percentage
caused_by                 141689           99.983770
vendor                    141468           99.827820
rfc                       140721           99.300694
problem_id                139417           98.380518
sys_created_by             53076           37.453427
sys_created_at             53076           37.453427
u_symptom                  32964           23.261262
assigned_to                27496           19.402732
assignment_group           14213           10.029496
opened_by                   4835            3.411849
resolved_at                 3141            2.216467
closed_code                  714            0.503839
resolved_by                  226            0.159478
subcategory                  111            0.078328
category                      78            0.055041
location                      76            0.053630
caller_id                     29            0.020464
contact_type                   0            0.

In [47]:
#analyzing first problematic column
df['caused_by'].head()

0    <NA>
1    <NA>
2    <NA>
3    <NA>
4    <NA>
Name: caused_by, dtype: object

notna() helps us to tell exactly how many records have a caused_by value

In [48]:
df['caused_by'].notna().sum()

23

Checking exactly at which number the caused_by has value through label based selection i.e df.loc

In [49]:
df.loc[df['caused_by'].notna(),['number','caused_by']]

,number,caused_by
1174,INC0000388,CHG0000132
1175,INC0000388,CHG0000132
1176,INC0000388,CHG0000132
1177,INC0000388,CHG0000132
1178,INC0000388,CHG0000132
1179,INC0000388,CHG0000132
1180,INC0000388,CHG0000132
4512,INC0001131,CHG0000097
4513,INC0001131,CHG0000097
4514,INC0001131,CHG0000097


Dropping caused_by -since the feature is extremely sparse ,making it unsuitable as a reliable predictive feature for the model

In [50]:
df.drop(columns=['caused_by'],inplace=True)

Analyzing vendor column before dropping

In [51]:
df['vendor'].head(5)

0    <NA>
1    <NA>
2    <NA>
3    <NA>
4    <NA>
Name: vendor, dtype: object

In [52]:
#printing non missing values
df['vendor'].notna().sum()

244

In [53]:
df.loc[df['vendor'].notna(),['number','vendor']]

,number,vendor
7857,INC0001845,code 8s
7858,INC0001845,code 8s
9086,INC0002099,code 8s
9087,INC0002099,code 8s
9088,INC0002099,code 8s
...,...,...
137209,INC0033686,Vendor 1
137210,INC0033686,Vendor 1
137211,INC0033686,Vendor 1
137212,INC0033686,Vendor 1


In [55]:
df['vendor'].nunique()

4

In [54]:
df['vendor'].value_counts()

vendor
code 8s     167
Vendor 1     69
Vendor 3      6
Vendor 2      2
Name: count, dtype: int64

In ITSM point of view a vendor is an external company/provider that supplies hardware,software,a product,service,or technical support to one organization

141,712 rows ≠ 141,712 incidents,as this is event log dataset,the same incident may appear multiple times.So we should check how many unique incidents have a vendor recorded.
 

In [60]:
df.loc[df['vendor'].notna(),'number'].unique()

array(['INC0001845', 'INC0002099', 'INC0002104', 'INC0002332',
       'INC0002336', 'INC0002559', 'INC0003216', 'INC0003228',
       'INC0003258', 'INC0004123', 'INC0005008', 'INC0005213',
       'INC0006165', 'INC0009186', 'INC0009537', 'INC0010272',
       'INC0010418', 'INC0011175', 'INC0011177', 'INC0011178',
       'INC0012044', 'INC0012984', 'INC0013219', 'INC0013636',
       'INC0013833', 'INC0013946', 'INC0015087', 'INC0015098',
       'INC0015283', 'INC0015300', 'INC0015847', 'INC0015927',
       'INC0015945', 'INC0015951', 'INC0015953', 'INC0016021',
       'INC0016565', 'INC0016783', 'INC0016889', 'INC0016942',
       'INC0017106', 'INC0019197', 'INC0019739', 'INC0021267',
       'INC0023726', 'INC0023739', 'INC0024173', 'INC0024227',
       'INC0024381', 'INC0024839', 'INC0025213', 'INC0025982',
       'INC0026235', 'INC0026474', 'INC0027949', 'INC0028354',
       'INC0028417', 'INC0029765', 'INC0029905', 'INC0029950',
       'INC0030004', 'INC0030649', 'INC0030708', 'INC00

Excluding vendor since vendor information is only available for 67 incidents out of 24918 unique incidents,indicating that vendor association is recorded only for very few incidents

In [61]:
df.drop(columns=['vendor'],inplace=True)